### PR TITLE
Use cyber.stake to resource limits #295

### PR DESCRIPTION
**Pull request by kaynarov**
this PR binds the user's stake to resource limits.

a user can stake sys tokens for 3 different purposes: CPU, NET, RAM.
the current share of the user's stake for a specific purpose determines which part of the total virtual limit of this resource the user can spend.

some notes:

* since users can proxy their tokens, and get the rewards (fines) of their agents, their proxied balances change over time. when updating balances (see update_proxied_traversal), we rely on the fact that rewards do not occur more than once per frame_length. the frequency with which rewards occur in a govern contract must match this value;

* EOS has a mechanism of pending limits after staking/unstaking. In the proposed version this mechanism is removed since lazy calculation is used.

* the user can spend more RAM than should according to the share. however, if the user has staked_virtual_balance < 0, the transaction will fail (see get_account_limit_ex and max_bandwidth_an_ccounts_can_pay)

### CODE REVIEW

File: libraries/chain/include/eosio/chain/resource_limits.hpp [hunk 1]
```diff
@@ -48,7 +46,12 @@ namespace resource_limits {
       int64_t used = 0; ///< quantity used in current window
       int64_t available = 0; ///< quantity available in current window (based upon fractional reserve)
       int64_t max = 0; ///< max per window under current congestion
+      int64_t staked_virtual_balance = 0;
```
Comment (kaynarov) [inline libraries/chain/include/eosio/chain/resource_limits.hpp]: propose to add an additional bandwidth limits for users with insufficient ram-staked tokens by calculating the virtual value of user's resources. for these users it is negative for ram, and >= 0 for bandwidths; total value can't be negative for a successful transaction.

File: libraries/chain/transaction_context.cpp [hunk 2]
```diff
@@ -579,28 +576,91 @@ namespace bacc = boost::accumulators;
...
       // Calculate the new highest network usage and CPU time that all of the billed accounts can afford to be billed
       auto& rl = control.get_mutable_resource_limits_manager();
       const static int64_t large_number_no_overflow = std::numeric_limits<int64_t>::max()/2;
       int64_t account_net_limit = large_number_no_overflow;
       int64_t account_cpu_limit = large_number_no_overflow;
+      int64_t block_time = control.pending_block_time().sec_since_epoch();
+      
+      std::map<account_name, int64_t> providers_virtual_balances;
+      if (check_staked_virtual_balance) {
+          for ( const auto& a : bill_to_accounts ) {
```
Comment (zxcat) [inline libraries/chain/transaction_context.cpp]: let's unify it with `if (some)` — without spaces around contents of parentheses

File: libraries/chain/resource_limits.cpp [hunk 3]
```diff
@@ -369,82 +212,73 @@ uint64_t resource_limits_manager::get_block_net_limit() const {
...
+    
+    symbol_code token_code { symbol(CORE_SYMBOL).to_symbol_code() };
+    
+    const stake_param_object* param = nullptr;
+    const stake_stat_object* stat = nullptr;
+    
+    if (_db.get<account_object, by_name>(account).privileged || //assignments:
+        !(param = _db.find<stake_param_object, by_id>(token_code.value)) || 
+        !(stat  = _db.find<stake_stat_object, stake_stat_object::by_key>(stat_key(purpose_code, token_code))) || 
+        !stat->enabled || stat->total_staked == 0) {
```
[the remark this answers is not in the document]
Comment (kaynarov) [reply libraries/chain/resource_limits.cpp]: we have to check that the contract is loaded and initialized, and there are staked tokens (the same logic was in EOS). I also added enabled flag, it is convenient for tests.

File: libraries/chain/resource_limits.cpp [hunk 4]
```diff
@@ -369,82 +212,73 @@ uint64_t resource_limits_manager::get_block_net_limit() const {
...
+    const auto& config = _db.get<resource_limits_config_object>();
+    const auto& state  = _db.get<resource_limits_state_object>();
+    const auto& usage  = _db.get<resource_usage_object, by_owner>(account);
+    int64_t capacity = 0;
+    int64_t used = 0;
+    if (purpose_code == cpu_code || purpose_code == net_code) {
+        bool cpu = (purpose_code == cpu_code);
+        int128_t window_size = cpu ? config.account_cpu_usage_average_window : config.account_net_usage_average_window;
+        
+        EOS_ASSERT(cpu ? state.virtual_cpu_limit : state.virtual_net_limit <= static_cast<uint64_t>(std::numeric_limits<int64_t>::max()),
```
[the remark this answers is not in the document]
Comment (kaynarov) [reply libraries/chain/resource_limits.cpp]: thx, we definitely need parentheses here

File: libraries/chain/resource_limits.cpp [hunk 5]
```diff
@@ -135,54 +135,13 @@ void resource_limits_manager::add_transaction_usage(const flat_set<account_name>
...
-      int64_t net_weight;
-      int64_t cpu_weight;
-      get_account_limits( a, unused, net_weight, cpu_weight );
 
       _db.modify( usage, [&]( auto& bu ){
           bu.net_usage.add( net_usage, time_slot, config.account_net_usage_average_window );
           bu.cpu_usage.add( cpu_usage, time_slot, config.account_cpu_usage_average_window );
       });
-
-      if( cpu_weight >= 0 && state.total_cpu_weight > 0 ) {
```
Comment (s-medvedev) [inline libraries/chain/resource_limits.cpp]: Why do you remove this checks?

**Pull request merged**